docs/decisions/0006-reorder-component-tree-to-reduce-unnecessary-api-calls-and-improve-user-experience.md

```diff
@@ -0,0 +1,103 @@
+# 0006. Eliminate unnecessary API calls and reduce bundle size
+
+## Context
+
+### Component Architecture
+
+Today, each page in the core enterprise learner experience within `frontend-app-learner-portal-enterprise` renders the `AuthenticatedUserSubsidyPage` component. For example, the "Dashboard" page is rendered via the `DashboardPage` component and similarly for `SearchPage`:
+
+```jsx
+// DashboardPage
+<AuthenticatedUserSubsidyPage>
+  <Dashboard />
+</AuthenticatedUserSubsidyPage>
+```
+
+```jsx
+// CoursePage
+<AuthenticatedUserSubsidyPage>
+  <Course />
+</AuthenticatedUserSubsidyPage>
+```
+
+The other core pages like the "Search", "Program", etc. also each independently render their own `AuthenticatedUserSubsidyPage` component as part of their route's logic. This means that, as a learner, when navigating from "Dashboard" to "Search", we are rendering `AuthenticatedUserSubsidyPage` (and performing all its logic) once and then again when the "Search" page loads, unnecessarily duplicating logic between page views.
+
+The logic encapsulated by `AuthenticatedUserSubsidyPage` entails:
+* Redirecting to logistration, via the enterprise `/proxy-login/` view in the LMS.
+* Refresh the authenticated user's JWT, if necessary.
+* Fetching an enterprise customer configuration.
+* Fetching a `CustomerAgreement`, subscription licenses, enterprise coupon codes (and soon enterprise offers).
+* Recompute custom brand styles ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/master/src/components/layout/Layout.jsx#L19)).
+
+Clearly, rendering `AuthenticatedUserSubsidyPage` separately for each page route duplicates a lot of work. Some of these tasks are blocking items, in that a spinner or other loading state must be shown to the user before we can proceed to render anything on the page since all this logic deals with determining what experience the learner may have throughout the app.
+
+As a result, we previously introduced client side API caching to eliminate the need to re-fetch data from the API that we don't expect to change frequently (e.g., an enterprise customer configuration), thus reducing the impact of blocking loading states between route changes.
+
+Even then, there is a brief, but noticeable flicker between page changes due to this component architecure that could be eliminated by moving some components around in the component tree such that the aforementioned logic is no longer executed unnecessarily again between route changes.
+
+### Bundle Size
+
+As this application has grown in size, it now supports 12 routes (at the time of this writing):
```

Comment:
It would be great to add this to the README for this repo!

Reply:
Oh wait, we already have it in the README - we just need to update the README :)